dask/dataframe/optimize.py

```diff
@@ -10,20 +10,21 @@
 
 
 def optimize(dsk, keys, **kwargs):
-
     if isinstance(dsk, HighLevelGraph):
         # Think about an API for this.
         flat_keys = list(core.flatten(keys))
         dsk = optimize_read_parquet_getitem(dsk, keys=flat_keys)
         dsk = optimize_blockwise(dsk, keys=flat_keys)
         dsk = fuse_roots(dsk, keys=flat_keys)
-
-    dsk = ensure_dict(dsk)
-
-    if isinstance(keys, list):
-        dsk, dependencies = cull(dsk, list(core.flatten(keys)))
+        dsk = dsk.cull(set(flat_keys))
+        dsk = ensure_dict(dsk)
+        dependencies = None
     else:
-        dsk, dependencies = cull(dsk, [keys])
+        dsk = ensure_dict(dsk)
+        if isinstance(keys, list):
+            dsk, dependencies = cull(dsk, list(core.flatten(keys)))
+        else:
+            dsk, dependencies = cull(dsk, [keys])
```

Comment:
Perhaps we should enforce that Dask Dataframes always have HighLevelGraphs?

Reply:
I think that is a great idea. We can start by converting `dsk` to a HLG if it isn't already?

Reply:
I think that we would handle this in the `DataFrame` constructor.  First, if we're handed a raw dict in our own code then we maybe want to err.  To respect downstream users though we probably do want to allow this.  And yes, I would just put it in a single-layered HLG.  It looks like we already do this in Dask array

```python

    def __new__(cls, dask, name, chunks, dtype=None, meta=None, shape=None):
        self = super(Array, cls).__new__(cls)
        assert isinstance(dask, Mapping)
        if not isinstance(dask, HighLevelGraph):
            dask = HighLevelGraph.from_collections(name, dask, dependencies=())
        self.dask = dask
```

Reply:
It looks like it always does that: https://github.com/dask/dask/blob/22224bb0f75aee3e04dc42ca006807e4d2fc4138/dask/dataframe/core.py#L288-L290